### DISCUSSION TITLE
ISP uses PPPoe, so cannot use port forwarding.  Please outline my alternatives.

### DISCUSSION THREAD

**Discussion by heyho74** [mcguirepr89/BirdNET-Pi#827] [General]
As per the title, I've just realised (after several hours of frustration) that I cannot use port forwarding with my ISP.  

In another discussion, I see VPN tunneling being recommended as a possible alternative.  I'm reasonably pi and Linux literate, and would like a brief outline of how I'd go about setting this up if it's possible.  Many thanks.

**Comment by gerdesj** [comment]
I've been a network technician for many years and can help you get this sorted.  You can use a VPN to get external access to your Pi but let's eliminate all other possibilities.

Does your ISP use carrier grade NAT by any chance?  Do you know what your external IP address is.  You need the WAN or internet address on your router.

**Reply by heyho74** [reply]
Thanks for your reply.

If I query my public IP with whatismyip or similar, it returns 80.189.228.183, but my router homepage shows the WAN IP as 192.168.88.107(PPPoe), which is obviously a private address.

To clarify, I receive my internet connection via a line of sight wireless transmitter system, hence the PPPoe connection.

I have in the past been assigned a static IP, but recently requested this was set back to dynamic as streaming services were detecting it as a VPN for some reason.

**Reply by gerdesj** [reply]
Yes, that is called CGNAT.  It's another attempt to keep IPv4 from running out of addresses.  So your ISP has a small pool of addresses and their router NATs that to private addresses (RFC1918).  Then your router NATs that to your RFC1918 range. 

So looking you up: https://bgp.he.net/ip/80.189.228.183 reveals BT as your real ISP.  Hi from South Somerset!  Who is in the middle ie who do you actually pay for service?  Roughly where are you?  I may be able to point you at another solution.

**Reply by heyho74** [reply]
Yes, I understand the end provider is BT, as I can see a Plusnet router which I believe to be at the end of the line of sight wireless link.  I'm in Kent, and pay a company called clear picture who facilitate the connection.

**Reply by gerdesj** [reply]
I've just had a look at their website.  They seem to have a focus on delivering service to hard to reach places.  You have some options but most involve serious cash!  

If you have a friend within around two miles with a fixed internet connection and fairly good line of sight then I'd consider doing your own wifi link - the APs for a point to point link eg Ubiquiti https://eu.store.ui.com/collections/operator-airmax-devices are about £100 a pop.  If they are within 100m then ethernet is an option.  30 mile links are about £600 per end plus pylons!   

Mulvad and the like can probably sell you an external IP via VPN but I have never tried it.  If there are no other reasonable options I might be persuaded to provide a link via VPN to a fellow BNP user with rather minimal guarantees!

**Reply by heyho74** [reply]
Unfortunately everyone around here is in the same boat.  The copper lines to the nearest BT exchange are over 5km away, so that results in a connection capable of 1 - 2mbps on a good day. When this wireless service became available it was a revelation.

As I've said, they have in the past assigned me a static IP address, but for some reason the channel 4 streaming service took a dislike to it and seemed to think it was a VPN.  I may request a static address again.

Still no sign of fibre coming to our village (the price you pay for living on the edge of a nature reserve I guess), so my next door neighbour has just ordered starlink.

**Reply by gerdesj** [reply]
I have some very rural customers here in Somerset.  It might be worth getting a quote for a proper link from BT or Talktalk or Virgin or Cityfibre - ie a leased line.  It is likely to be something like £30,000.  That will get you the possibility of multiple 1Gb+ links.  They cost rather less than they used to say about £400 a month.  Now that's a bit mad for a single house but for an entire village it starts to look feasible.  A mesh wifi setup is say £100 a house for hardware and the cost of installation - reasonable DIY skills only needed.  If you can drill a hole in brick without hitting a cable then you are good.  You'll also need a central point of distribution with an AP costing more like £1000.

Say there are 100 houses in your village.  50% go for it.  That's roughly £1000 up front each + install costs and then an ongoing cost of not much plus someone to manage that lot.  Let's allow a bit of slack:  Say £1500 up front and say £300-£500 pa per household.

That gets you a sub 10ms latency link (allowing for your wifi link) - the leased line(s) will be sub 3ms and a decent share of 1 or more GB/s.

Starlink has rubbish latency but quite decent thoughput.  Rubbish for gaming, Fine for streaming and regular browsing.  Rubbish for hosting - highly asymmetric link speeds.

If you want to look into a village scale DIY internet then it isn't too mad.  You may collectively want to try negotiating with your current supplier for better terms.  

I'd go back to a static IP if you can.  Sod C4!  IP addresses have a "reputation" and you can gradually improve the reputation of yours by behaving "correctly".  I can send email from my home range and according to the experts half my age that should be impossible.

A few weeks ago me and the wife stayed in a hotel in the Algarve, Portugal near Alvor.  There's a reserve there covering the transition from the Atlantic, across the beaches etc to a lagoon.  There are loads of birds and other wildlife and it seems to strike a fair balance between the rampant tourism there and nature.  One of the birds on the information boards was called in English: the Kentish Plover.  I looked it up on Wikipedia and noted that it hasn't bred on our shores since 1979.  They were having a whale of a time down south, browning their knees in the sun and mincing in and out of the gentle waves on the lagoon.

**Reply by heyho74** [reply]
At last count I believe there were 26 properties in the village.  I suspect 20% of those wouldn't be interested in a high speed connection, which leaves approximately 20 properties to share the cost.  There was talk of a rural broadband grant,(150k for fibre) but I'm told we were fobbed off several times, and the contact at Openreach is now ignoring all correspondence.  To be honest, the 15mbps which we now receive is adequate for our needs, it's just the lack of a public IP that's annoying.

**Comment by ampsu** [comment]
https://ngrok.com/ - setting up is easy. Install it, add authtoken to config, open config file and add a port 80 http tunnel. Then install a service and start it.

**Reply by gerdesj** [reply]
https://ngrok.com/pricing - looks pretty reasonable and will almost certainly do the job as specified. 

Despite my wall of words above - I'd go with this, given your situation.

**Reply by heyho74** [reply]
Thanks to everyone who has suggested possible fixes.  I'll look into this when I get some time over the weekend.

**Reply by heyho74** [reply]
So, it turns out ngrok really was that simple.  Up and running in minutes.

One final question...  Is it possible to keep the ngrok tunnel/session open, without remaining connected via SSH on my LAN?

**Reply by ampsu** [reply]
Yes it is: https://ngrok.com/docs/secure-tunnels/ngrok-agent/installing-as-a-service/ 
You need to edit the ngrok.yml conf file and add: `"tunnels:
  website:
    addr: 80
    proto: http"` Or whatever port you use. I am not sure whether you need to specify protocol, I just left it there.
Then you need to install a service with that config. `ngrok service install --config path-to-youconfig`. You can get a path with `ngrok config check`. Then start config with `ngrok service start`. And most likely you need to run those commands with sudo, depending where did you install ngrok.

**Reply by heyho74** [reply]
All of these commands complete successfully, but I'm not getting a tunnel opening, and any trace of ngrok actually running (no process visible using "top") by either starting the service, or rebooting the pi.

I'm wondering if maybe it's not being pointed to the ngrok binary correctly?  I'm slightly out of my depth here, but enjoy learning.

**Reply by heyho74** [reply]
Okay, brief update for anybody else who goes down this route.  The format of the ngrok.yml conf file is very important with regard to layout and indents.  Don't ask me exactly what the layout should be, however defining the tunnel on one line doesn't work.  I ended up with the following:

tunnels:
  website:
    addr: 80
    proto: http

Which is now working.

Edit:  My indents are not showing...  there are 2 spaces before line 2, and 4 spaces before lines 3 and 4.

**Comment by matthew73210** [comment]
Zerotier, that's what I use. Not port forwarding, no internet facing webpage.
